File: api/src/services/auth.service.ts
import pool from "../db";
import { IUser } from "../interfaces/user.interface";
import helper from "../utils/helper";

/**
 * signUp()
 * @param {IUser} userParams
 * @returns {Promise<any>}
 */
const signUp = async (userPrams : IUser) : Promise<any> => {

    const {
        firstName,
        lastName,
        email,
        password,
        gender,
        status,
        phoneNumber,
        displayPhoto
    } = userPrams;

    const result = await pool.query(`
        INSERT INTO person (
            first_name,
            last_name,
            email,
            password,
            gender,
            status,
            phone_number,
            display_photo,
            created_by
        ) VALUES (
            $1,
            $2,
            $3,
            $4,
            $5,
            $6,
            $7,
            $8,
            lastval()
        ) RETURNING *;
    `,
    [firstName, lastName, email, password, gender, status, phoneNumber, displayPhoto])
    .catch(e => {throw e});

    return helper.keysToCamel(result);
};

/**
 * getCredentialsByEmail()
 * @param {string} email
 * @returns {Promise<any>}
 */
 const getCredentialsByEmail = async (email : string) : Promise<any> => {

    const result = await pool.query(`
        SELECT 
            person_id,
            email,
            password
        FROM person 
        WHERE email = $1 
        ORDER BY person_id ASC`, 
    [email])
    .catch(e => {throw e});

    return helper.keysToCamel(result.rows);
};

/**
 * updateRefreshToken()
 * @param {string} refreshToken
 * @param {string} id
 * @returns {Promise<any>}
 */
 const updateRefreshToken = async (refreshToken : string, id : string) : Promise<any> => {

    const result = await pool.query(`
        UPDATE person
        SET
            refresh_token = $1
        WHERE person_id = $2
        RETURNING *;
    `, 
    [refreshToken,id])
    .catch(e => {throw e});

    return helper.keysToCamel(result.rows);
};

export default {
    signUp,
    getCredentialsByEmail,
    updateRefreshToken
};

File: go_api/services/auth_service.go
package services

import (
	"fmt"

	"github.com/edsycarreon/pet/db"
	"github.com/edsycarreon/pet/models"
	"github.com/edsycarreon/pet/utils"
)

// SignUp()
func SignUp(person models.Person) (models.Person, error) {

	var p models.Person
	
	// Prepare SQL statement
	sql := `
		INSERT INTO person (
			first_name,
			last_name,
			email,
			password,
			gender,
			status,
			phone_number,
			display_photo,
			created_by
		) VALUES (
			:first_name,
			:last_name,
			:email,
			:password,
			:gender,
			'PEN',
			:phone_number,
			:display_photo,
			lastval()
		) RETURNING *;
	`

	res, err := db.DB.NamedQuery(sql, person)
	if err != nil {
		return p, err
	}

	if res.Next() {
		res.Scan(&p)
	}

	return p, nil
}

// GetCredentialsByEmail
func GetCredentialsByEmail(email string) (models.Person, error) {

	var p models.Person

	// Prepare SQL statement
	sql := `
		SELECT 
			person_id,
			email,
			password
		FROM person 
		WHERE email = $1 
		ORDER BY person_id ASC
	`

	err := db.DB.Get(&p, sql, email)
	if err != nil {
		return p, err
	}

	utils.PrintJSON(p)

	return p, nil
}

// ChangePassword
func ChangePassword(cred models.Person) (models.Person, error) {

	var p models.Person	

	fmt.Println(cred)
	// Prepare sql statement
	sql := `
		UPDATE person
		SET password = :new_password
		WHERE person_id = :person_id
	`

	res, err := db.DB.NamedQuery(sql, cred)
	if err != nil {
		return p, err
	}

	if res.Next() {
		res.Scan(&p)
	}

	return p, nil
}